src/ExpressionEvaluator/Core/Source/ResultProvider/ResultProvider.cs

```diff
@@ -390,52 +514,109 @@ internal DkmEvaluationResult GetResult(EvalResultDataItem dataItem, DkmClrType r
+                GetEvaluationResultsAndContinue(rows, initialChildren, 0, numRows, workList, inspectionContext, stackFrame,
+                    () => ContinueWithExceptionHandling(
+                        () =>
+                        {
+                            var enumContext = DkmEvaluationResultEnumContext.Create(index, stackFrame, inspectionContext, new EnumContextDataItem(dataItem));
+                            completionRoutine(new DkmGetChildrenAsyncResult(initialChildren, enumContext));
+                        },
+                        e => completionRoutine(DkmGetChildrenAsyncResult.CreateErrorResult(e))));
+            }
+            finally
             {
-                expansion.GetRows(this, builder, inspectionContext, dataItem, dataItem.Value, 0, initialRequestSize, visitAll: true, index: ref index);
+                rows.Free();
             }
-            var rows = builder.ToArrayAndFree();
-            Debug.Assert(index >= rows.Length);
-            Debug.Assert(initialRequestSize >= rows.Length);
-            var enumContext = DkmEvaluationResultEnumContext.Create(index, evaluationResult.StackFrame, evaluationResult.InspectionContext, new EnumContextDataItem(dataItem));
-            return new DkmGetChildrenAsyncResult(rows, enumContext);
         }
 
-        private DkmEvaluationEnumAsyncResult GetItems(DkmInspectionContext inspectionContext, EvalResultDataItem dataItem, int startIndex, int count)
+        private void GetItemsAndContinue(EvalResultDataItem dataItem, DkmWorkList workList, int startIndex, int count, DkmInspectionContext inspectionContext, DkmCompletionRoutine<DkmEvaluationEnumAsyncResult> completionRoutine)
         {
             var expansion = dataItem.Expansion;
-            var builder = ArrayBuilder<DkmEvaluationResult>.GetInstance();
-            if (expansion != null)
+            var value = dataItem.Value;
+            var rows = ArrayBuilder<EvalResultDataItem>.GetInstance();
+            try
             {
-                int index = 0;
-                expansion.GetRows(this, builder, inspectionContext, dataItem, dataItem.Value, startIndex, count, visitAll: false, index: ref index);
+                if (expansion != null)
+                {
+                    int index = 0;
+                    expansion.GetRows(this, rows, inspectionContext, dataItem, value, startIndex, count, visitAll: false, index: ref index);
+                }
+                var numRows = rows.Count;
+                Debug.Assert(count >= numRows);
+                var results = new DkmEvaluationResult[numRows];
+                GetEvaluationResultsAndContinue(rows, results, 0, numRows, workList, inspectionContext, value.StackFrame,
+                    () => ContinueWithExceptionHandling(
+                        () => completionRoutine(new DkmEvaluationEnumAsyncResult(results)),
+                        e => completionRoutine(DkmEvaluationEnumAsyncResult.CreateErrorResult(e))));
+            }
```

Comment:
Is it necessary to wrap the calls in ContinueWithExceptionHandling? If completionRoutine already handles exceptions, would it be sufficient to simply call completionRoutine(new DkmEvaluationEnumAsyncResult(results))?